### PR TITLE
case macro now can be used inside generic. Fixes #20435

### LINKED ISSUE #20435: Compilation fails when using fusion's 'Some' inside a generic proc
### What happened?

The following code will fail to compile due to willFail being generic

```nim

import std/options
import fusion/matching


proc works() = 
  let a = some 2
  case a:
  of Some(@t):
    echo "some"
  of None():
    echo "none"

proc willFail[T]() = 
  let b = some 2
  case b:
  of Some(@t):
    echo "some"
  of None():
    echo "none"

```

### Nim Version

```
Nim Compiler Version 1.7.1 [MacOSX: amd64]
Compiled at 2022-09-13
Copyright (c) 2006-2022 by Andreas Rumpf
```

### Current Standard Output Logs

```shell
Error: undeclared identifier: 'Some'

```
```


### Expected Standard Output Logs

```shell
[SuccessX]

```
```


### Possible Solution

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by jmgomez**
fixes #20435

### CODE REVIEW

File: compiler/semgnrc.nim [hunk 1]
```diff
@@ -356,13 +360,25 @@ proc semGenericStmt(c: PContext, n: PNode,
   of nkCaseStmt:
     openScope(c)
     n[0] = semGenericStmt(c, n[0], flags, ctx)
+    var errors: CandidateErrors
+    let r = resolveCaseSmtMacroOverload(c, n, errors)
```
Comment (Araq) [inline compiler/semgnrc.nim]: Too early. Instead you need to special case `of` branches to allow for unknown identifiers.

Comment (jmgomez) [reply compiler/semgnrc.nim]: There are two parts of this:
1. To allow the syntax. My attempt tries to see if the `case(exp)` matches the macro. 
2. The macro needs to be expanded so you account for the `@` identifiers (`a`, `b`) below
```nim
  proc regularCase[T]() = 
    let foo = [(1, 3), (3, 4)]
    case foo:
    of [(1, @a), (_, @b)]:
      echo a, b
    else: discard
    
```
Your comment is in reference to [1] or to both?

**Comment by ringabout** [comment]
Hello, I have edited the header to include `Fixes #20435`. It's a convention that helps reviewers quickly navigate to the specific issue and etc.

File: compiler/semgnrc.nim [hunk 2]
```diff
@@ -360,8 +360,8 @@ proc semGenericStmt(c: PContext, n: PNode,
       var a = n[i]
       checkMinSonsLen(a, 1, c.config)
       for j in 0..<a.len-1:
-        a[j] = semGenericStmt(c, a[j], flags, ctx)
-      a[^1] = semGenericStmtScope(c, a[^1], flags, ctx)
+        a[j] = semGenericStmt(c, a[j], flags + {withinMixin}, ctx)
```
Comment (Araq) [inline compiler/semgnrc.nim]: This should use `addTempDecl(c, s, skVar)` for nkIdents in the `of` branch.

File: compiler/semgnrc.nim [hunk 3]
```diff
@@ -360,8 +360,8 @@ proc semGenericStmt(c: PContext, n: PNode,
...
+      a[^1] = semGenericStmtScope(c, a[^1], flags + {withinMixin}, ctx)
```
Comment (Araq) [inline compiler/semgnrc.nim]: And then for the body it doesn't have to use `withinMixin`

File: compiler/semgnrc.nim [hunk 4]
```diff
@@ -360,7 +367,9 @@ proc semGenericStmt(c: PContext, n: PNode,
       var a = n[i]
       checkMinSonsLen(a, 1, c.config)
       for j in 0..<a.len-1:
-        a[j] = semGenericStmt(c, a[j], flags, ctx)
+        a[j] = semGenericStmt(c, a[j], flags + {withinMixin}, ctx)
+        addTempDeclToIdents(c, a[j], skVar)
```
Comment (Araq) [inline compiler/semgnrc.nim]:
```suggestion
        addTempDeclToIdents(c, a[j], skVar)
        a[j] = semGenericStmt(c, a[j], flags, ctx)
```

File: compiler/semgnrc.nim [hunk 5]
```diff
@@ -360,7 +367,9 @@ proc semGenericStmt(c: PContext, n: PNode,
       var a = n[i]
       checkMinSonsLen(a, 1, c.config)
       for j in 0..<a.len-1:
-        a[j] = semGenericStmt(c, a[j], flags, ctx)
+        a[j] = semGenericStmt(c, a[j], flags+{withinMixin}, ctx)
+        addTempDeclToIdents(c, a[j], skVar)
```
Comment (Araq) [inline compiler/semgnrc.nim]:
```suggestion
        addTempDeclToIdents(c, a[j], skVar, false)
```

File: compiler/semgnrc.nim [hunk 6]
```diff
@@ -188,6 +188,13 @@ proc addTempDecl(c: PContext; n: PNode; kind: TSymKind) =
   styleCheckDef(c, n.info, s, kind)
   onDef(n.info, s)
 
+proc addTempDeclToIdents(c: PContext; n: PNode; kind: TSymKind) =
+  if n.kind == nkIdent:
+    addTempDecl(c, n, kind)
+  else:
+    for s in n:
+      addTempDeclToIdents(c, s, kind)
+
```
Comment (Araq) [inline compiler/semgnrc.nim]:
```suggestion
proc addTempDeclToIdents(c: PContext; n: PNode; kind: TSymKind; inCall: bool) =
  case n.kind 
  of nkIdent:
    if inCall:
      addTempDecl(c, n, kind)
  of nkCallKinds:
    for s in n:
      addTempDeclToIdents(c, s, kind, true)  
  else:
    for s in n:
      addTempDeclToIdents(c, s, kind, inCall)

```

**Pull request merged**